packages/environment-ember-template-imports/-private/environment/preprocess.ts

```diff
@@ -1,54 +1,91 @@
-    (match) => match.type === 'template-tag'
-  );
+  // NOTE: https://github.com/embroider-build/content-tag/issues/45
+  //       All indicies are byte-index, not char-index.
+  let templates = p.parse(source, path);
 
   let templateLocations: Array<TemplateLocation> = [];
   let segments: Array<string> = [];
-  let sourceOffset = 0;
-  let delta = 0;
+  let sourceOffsetBytes = 0;
+  let deltaBytes = 0;
+
+  // @ts-expect-error TS couldn't find @types/node, which are specified
+  // in the root package.json
+  let sourceBuffer = Buffer.from(source);
 
   for (let template of templates) {
-    let { start, end } = template;
-    let startTagLength = template.start[0].length;
-    let endTagLength = template.end[0].length;
-    let startTagOffset = start.index ?? -1;
-    let endTagOffset = end.index ?? -1;
+    let startTagLengthBytes = template.startRange.end - template.startRange.start;
+    let endTagLengthBytes = template.endRange.end - template.endRange.start;
+    let startTagOffsetBytes = template.startRange.start;
+    let endTagOffsetBytes = template.endRange.start;
 
-    if (startTagOffset === -1 || endTagOffset === -1) continue;
+    // if (startTagOffset === -1 || endTagOffset === -1) continue;
 
-    let transformedStart = startTagOffset - delta;
+    let transformedStartBytes = startTagOffsetBytes - deltaBytes;
 
-    segments.push(source.slice(sourceOffset, startTagOffset));
+    /**
+     * TODO: we want content-tag to manage all this for us, as managing indicies
+     *       can be error-prone.
+     *
+     * SEE: https://github.com/embroider-build/content-tag/issues/39#issuecomment-1832443310
+     */
+    let prefixingSegment = sourceBuffer.slice(sourceOffsetBytes, startTagOffsetBytes);
+    segments.push(prefixingSegment.toString());
     segments.push(TEMPLATE_START);
-    delta += startTagLength - TEMPLATE_START.length;
 
-    let transformedEnd = endTagOffset - delta + TEMPLATE_END.length;
+    // For TEMPLATE_START & TEMPLATE_END, characters === bytes
+    deltaBytes += startTagLengthBytes - TEMPLATE_START.length;
+
+    let transformedEnd = endTagOffsetBytes - deltaBytes + TEMPLATE_END.length;
 
-    segments.push(source.slice(startTagOffset + startTagLength, endTagOffset));
+    let templateContentSegment = sourceBuffer.slice(
+      startTagOffsetBytes + startTagLengthBytes,
+      endTagOffsetBytes
+    );
+    segments.push(templateContentSegment.toString());
     segments.push(TEMPLATE_END);
-    delta += endTagLength - TEMPLATE_END.length;
+    deltaBytes += endTagLengthBytes - TEMPLATE_END.length;
 
-    sourceOffset = endTagOffset + endTagLength;
+    sourceOffsetBytes = endTagOffsetBytes + endTagLengthBytes;
 
+    // TODO: is there a way to convert bytes to chars?
+    //       I think maybe all of this code needs to live in content-tag,
+    //       and give us the option to generate this sort of structure
     templateLocations.push({
```

Comment:
This is all likely wrong, as the original code used character-indices

Reply:
Is this still likely wrong?

Reply:
yeah, char positions are expected, but I only have bytes here

Reply:
This is how it is currently done in eslint parser:
https://github.com/NullVoxPopuli/ember-eslint-parser/blob/d1b8796436670e496f4a21de8b1a9d9219038b66/src/parser/transforms.js#L25

Reply:
ah ok, I'll try that

Reply:
pushed!

this is looking excellent!

![image](https://github.com/typed-ember/glint/assets/199018/403bccd5-4838-4208-b5bd-0d3fa0a87663)
![image](https://github.com/typed-ember/glint/assets/199018/aa8ff147-a7f2-43e1-8e65-fe048ac3bd9e)
![image](https://github.com/typed-ember/glint/assets/199018/8a31a8d8-3470-40ee-874c-24376658a4bc)